docs/lensvm/lens-vm.md

```diff
@@ -0,0 +1,161 @@
+
+## Before you begin
+
+Before getting started, ensure the following are installed:
+
+- [Golang](https://golang.google.cn/doc/install) (required to run the Lens engine)
+
+- WASM compiler (for your language of choice)
+  - You’ll need a compiler that targets the `wasm32-unknown` architecture.
+
+**Note**: A LensVM Engine is the runtime that executes Lenses in isolated WebAssembly environments, managing data flow, memory, and function calls between the host and the Lens.
+
+## Writing lenses
+
+Lenses can be written in any language that compiles to valid WebAssembly (WASM). The LensVM engine interacts with the compiled Lens module via a specific set of exported and imported functions.
+
+### Required and optional functions
+
+Each Lens must implement the following interface:
+
+| Function               | Type      | Required | Description |
+|:-----------------------|:----------|:---------|:------------|
+| `alloc(unsigned64)`    | Exported  | Yes        | Allocates a memory block of the given size. Called by the LensVM engine. |
+| `next() -> unsigned8`  | Imported  | Yes        | Called by the Lens to retrieve a pointer to the next input data item from the engine. |
+| `set_param(unsigned8) -> unsigned8` | Exported | No | Accepts static configuration data at initialization. Receives a pointer to the config and returns a pointer to an OK or error response. Called once before any input is processed. |
+| `transform() -> unsigned8` | Exported | Yes     | Core transformation logic. Pulls zero or more inputs using `next()`, applies transformation, and returns a pointer to a single output item. Supports stateful transformations. |
+| `inverse() -> unsigned8`   | Exported | No     | Optional reverse transformation logic, same interface as `transform()`. |
+
+### WASM data format
+
+Data passed across the WASM boundary—into `set_param()`, `transform()`, and `inverse()`—follows this binary format:
+
+```json
+
+[TypeId][Length][Payload]
+```
+
+- **TypeId**: Signed 8-byte integer  
+- **Length**: Unsigned 32-byte integer (optional depending on TypeId)  
+- **Payload**: Raw binary or serialized data (e.g., JSON)
+
+#### TypeId Values
+
+| TypeId | Meaning            | Notes |
+|:-------|:--------------------|:------|
+| `-1`   | Error               | May include an error message in the Payload. |
+| `0`    | Nil                | No `Length` or `Payload`. |
+| `1`    | JSON               | Payload contains a JSON-serialized object. |
+| `127`  | End of Stream      | Indicates there are no more items to process. |
+
+### Rust SDK
+
+To simplify Lens development, a [Rust SDK](https://docs.rs/lens_sdk) is available. It handles memory allocation, pointer management, and payload formatting, so you can focus on writing transformation logic.
```

Comment:
suggestion: `It handles memory allocation, pointer management, and payload formatting` is not quite accurate and makes writing lenses look harder than it actually is - it handles those things for the transfer of data across the wasm boundary, as well as the boilerplate required for meet the interface requirements.

It might be better to change this to something like: `It handles the boilerplate required to fullfill the above interfaces and safely transfer data across the wasm boundary, so you can focus on writing transformation logic.`